src/staker.cairo

```diff
@@ -212,8 +232,104 @@ pub mod Staker {
                 self.find_delegated_cumulative(delegate, mid, max_index_exclusive, timestamp)
             }
         }
+
+        fn log_change(ref self: ContractState, amount: u128, is_add: bool) {
+            let log = self.staking_log.as_path();
+
+            if log.len() == 0 {
+                // Add the first record. If withdrawal, then it's underflow.
+                assert(is_add, 'BAD AMOUNT'); 
+
+                log.append().write(
+                    StakingLogRecord {
+                        timestamp: get_block_timestamp(),
+                        total_staked: amount,
+                        cumulative_seconds_per_total_staked: 0_u64.into(),
+                    }
+                );
+                
+                return;
+            }
+
+            let last_record_ptr = log.at(log.len() - 1);
+                
+            let mut last_record = last_record_ptr.read();
+
+            let mut record = if last_record.timestamp == get_block_timestamp() {
+                // update record
+                last_record_ptr 
+            } else {
+                // create new record
+                log.append()
+            };
+
+            // Might be zero
+            let seconds_diff = (get_block_timestamp() - last_record.timestamp) / 1000;
+                
+            let staked_seconds: UFixedPoint = if last_record.total_staked == 0 {
+                0_u64.into()
+            } else {
+                seconds_diff.into() / last_record.total_staked.into()
+            };
+
+            let total_staked = if is_add {
+                // overflow check
+                assert(last_record.total_staked + amount >= last_record.total_staked, 'BAD AMOUNT'); 
+                last_record.total_staked + amount
+            } else {
+                // underflow check
+                assert(last_record.total_staked >= amount, 'BAD AMOUNT'); 
+                last_record.total_staked - amount
+            };
```

Comment:
if you only store cumulatives you can avoid this write in case where the timestamp is not changed

Reply:
Makes sence.